File: src/lib/sabflow/queue/rate-limit.ts
/**
 * SabFlow queue rate-limit — per-workspace + per-plan throttle.
 *
 * Track B Phase 2 / sub-task #7: protect the SabFlow execution queue from
 * runaway tenants by capping `workflow_executions/minute` at enqueue time.
 * The Rust dispatcher (`rust/crates/sabflow-executor/queue/src/rate_limit.rs`)
 * runs the same check at claim time against the same Redis keys, so a burst
 * that slips past the API boundary is still refused before a worker picks
 * it up.
 *
 * ## Algorithm
 *
 * Fixed-bucket-per-minute sliding window:
 *
 *   - Key: `sabflow:rate:<plan>:<workspaceId>:<minute>` where `<minute>` is
 *     `Math.floor(Date.now() / 60_000)`.
 *   - On each check: INCR the key, then EXPIRE 65s so old buckets evict
 *     themselves. (65s = 60s window + 5s slack so a check that lands
 *     exactly on a minute boundary still sees a fresh TTL.)
 *   - If the post-INCR value exceeds the plan cap, the call is denied and
 *     we decrement back so the counter reflects accepted work only.
 *
 * The "sliding window counter" name is a slight misnomer — we use a fixed
 * bucket per wall-clock minute, which is the standard cheap approximation.
 * A request right at the minute boundary could in theory burst at 2×cap;
 * the dispatcher-side check is the safety net for that edge.
 *
 * ## Plan caps (`workflow_executions/minute`)
 *
 *   free       → 5
 *   starter    → 30
 *   pro        → 120
 *   business   → 600
 *   enterprise → unlimited (no Redis round-trip; we short-circuit)
 *
 * ## Usage-meter integration
 *
 * When a request is allowed, we forward to the billing meter via
 * `recordUsage('workflow_executions', workspaceId, 1)`. This records queue
 * submission, not successful runs. The meter's own idempotency-key support
 * is the right place to decide whether failed runs should be refunded; we
 * intentionally don't gate the meter call on success here.
 */

import 'server-only';
import { recordUsage } from '@/lib/billing/usage-meter';

// ── Plan caps ─────────────────────────────────────────────────────────

/** Plan IDs recognised by this limiter. Extra plans fall through to `free`. */
export type PlanId = 'free' | 'starter' | 'pro' | 'business' | 'enterprise';

/** `Infinity` means "unlimited — skip the Redis check entirely". */
const PLAN_CAPS_PER_MINUTE: Record<PlanId, number> = {
    free: 5,
    starter: 30,
    pro: 120,
    business: 600,
    enterprise: Infinity,
};

/** Bucket TTL: 60s window + 5s slack. */
const BUCKET_TTL_SEC = 65;

// ── Public surface ────────────────────────────────────────────────────

export interface EnqueueRateCheckInput {
    workspaceId: string;
    plan: PlanId | string;
}

export interface EnqueueRateCheckResult {
    /** True when the caller may proceed to enqueue. */
    allowed: boolean;
    /** Tokens left in the current minute window after this check. */
    remaining: number;
    /** Seconds the caller should wait before retrying. Always >= 1 on deny. */
    retryAfterSec: number;
}

/**
 * Check (and atomically reserve) one slot in the per-workspace+plan minute
 * bucket. When `allowed === true` we also record one
 * `workflow_executions` usage event so the billing meter reflects queue
 * submission.
 *
 * Failure modes:
 *   - Unknown plan id → treated as `free` (the strictest cap).
 *   - Redis unavailable → fail-open (allow + log warning); the dispatcher's
 *     claim-side check is the secondary line of defense. Throttling is a
 *     guard rail, not a correctness invariant, so a degraded Redis must not
 *     make the platform unusable.
 */
export async function checkEnqueueRate(
    input: EnqueueRateCheckInput,
): Promise<EnqueueRateCheckResult> {
    if (!input.workspaceId) {
        throw new Error('workspaceId required');
    }

    const plan = normalisePlan(input.plan);
    const cap = PLAN_CAPS_PER_MINUTE[plan];

    // Unlimited plans skip both the Redis hit and the meter write — the
    // latter is recorded by the caller after the job lands in the queue
    // (idempotency-keyed on the executionId). Surfacing `remaining` as
    // `Number.MAX_SAFE_INTEGER` is a deliberate signal to UI code that
    // "don't bother showing a quota indicator".
    if (!Number.isFinite(cap)) {
        return { allowed: true, remaining: Number.MAX_SAFE_INTEGER, retryAfterSec: 0 };
    }

    const minute = Math.floor(Date.now() / 60_000);
    const key = bucketKey(plan, input.workspaceId, minute);

    let client: RedisLike | null = null;
    try {
        client = await getRedis();
    } catch (err) {
        // Fail-open. The dispatcher will still re-check at claim time and
        // operators get a `redis_unavailable` warning in the logs.
        // eslint-disable-next-line no-console
        console.warn('[sabflow.rate-limit] redis unavailable, failing open', err);
        await recordUsage({
            tenantId: input.workspaceId,
            feature: 'workflow_executions',
            units: 1,
        }).catch(() => undefined);
        return { allowed: true, remaining: cap, retryAfterSec: 0 };
    }

    const count = await client.incr(key);
    if (count === 1) {
        // First hit in this minute — set TTL. We always re-EXPIRE on the
        // first INCR rather than every call to keep the round-trip small.
        await client.expire(key, BUCKET_TTL_SEC);
    }

    if (count > cap) {
        // Roll back so the counter only reflects accepted work. A best-effort
        // DECR is fine here — if it fails we'll just over-count for one
        // minute and the bucket evicts itself.
        await client.decr(key).catch(() => undefined);
        const retryAfterSec = secondsUntilNextMinute();
        return { allowed: false, remaining: 0, retryAfterSec };
    }

    // Allowed — write the meter event before returning. We don't await the
    // meter write blocking the queue submission would be silly; the meter
    // is append-only and tolerant of duplicates via its idempotency key.
    void recordUsage({
        tenantId: input.workspaceId,
        feature: 'workflow_executions',
        units: 1,
    }).catch((err) => {
        // eslint-disable-next-line no-console
        console.warn('[sabflow.rate-limit] usage-meter write failed', err);
    });

    return {
        allowed: true,
        remaining: Math.max(0, cap - count),
        retryAfterSec: 0,
    };
}

// ── Internals ─────────────────────────────────────────────────────────

/**
 * Coerce an arbitrary plan string to a known PlanId. Unknown values fall
 * to `free` — the strictest cap — so a misconfigured tenant can never
 * accidentally unlock a higher rate by sending garbage.
 */
function normalisePlan(plan: string): PlanId {
    if (plan in PLAN_CAPS_PER_MINUTE) {
        return plan as PlanId;
    }
    return 'free';
}

/**
 * Produce the Redis key for a bucket. Exported via `_internals` for the
 * dispatcher-side and tests; not part of the stable public surface.
 */
function bucketKey(plan: PlanId, workspaceId: string, minute: number): string {
    return `sabflow:rate:${plan}:${workspaceId}:${minute}`;
}

function secondsUntilNextMinute(): number {
    const remainderMs = 60_000 - (Date.now() % 60_000);
    return Math.max(1, Math.ceil(remainderMs / 1000));
}

/** Minimal shape we need from the Redis client. */
interface RedisLike {
    incr(key: string): Promise<number>;
    decr(key: string): Promise<number>;
    expire(key: string, seconds: number): Promise<unknown>;
}

/**
 * Lazy ioredis client — matches the convention in
 * `src/lib/sabflow/recipes/durable.ts`. We keep the singleton on
 * `globalThis` so HMR / multiple imports don't open extra connections.
 */
declare global {
    // eslint-disable-next-line no-var
    var __sabflowRateLimitRedis: RedisLike | undefined;
}

async function getRedis(): Promise<RedisLike> {
    if (globalThis.__sabflowRateLimitRedis) return globalThis.__sabflowRateLimitRedis;
    const { default: IORedis } = await import('ioredis');
    const url = process.env.REDIS_URL || 'redis://localhost:6379';
    const client = new IORedis(url, {
        maxRetriesPerRequest: 2,
        enableReadyCheck: true,
        lazyConnect: false,
    }) as unknown as RedisLike;
    globalThis.__sabflowRateLimitRedis = client;
    return client;
}

/** Test/maintenance hooks. Not part of the stable public surface. */
export const _internals = {
    bucketKey,
    secondsUntilNextMinute,
    PLAN_CAPS_PER_MINUTE,
    BUCKET_TTL_SEC,
};

File: rust/crates/sabflow-executor/queue/src/rate_limit.rs
//! Claim-side per-workspace + per-plan rate limiter.
//!
//! Mirrors the TS enqueue-side check in
//! `src/lib/sabflow/queue/rate-limit.ts`. The two helpers share the same
//! Redis key shape so a request that was counted at enqueue time is the
//! same one we observe here — the dispatcher just re-runs the check
//! defensively before it claims a job.
//!
//! ## Algorithm
//!
//! Fixed bucket per wall-clock minute. INCR + EXPIRE 65s. If the post-INCR
//! count exceeds the plan cap we DECR back and return [`RateCheck::Denied`]
//! with `retry_after_ms` set to the time until the next minute boundary.
//!
//! ## Why not Lua?
//!
//! INCR is already atomic against a single key, and our EXPIRE is only set
//! on the first hit (when `count == 1`). The only race is "two callers
//! INCR the same key from 0 to 2 in the same Redis tick and both think
//! they were first"; both will then SET the same TTL on the same key. That
//! is a no-op, not a correctness bug — so a Lua round-trip is unnecessary
//! and we keep the dependency surface (`fred` without `i-scripts`) smaller
//! than the broadcast token-bucket crate.
//!
//! ## Failure mode
//!
//! Redis transport errors are returned as `anyhow::Error` from
//! `[`check_claim_rate`]`. The dispatcher's policy is to *fail closed* on
//! the claim side — if we can't tell whether the workspace is throttled,
//! we refuse the claim and let the next tick try again. That's the
//! opposite of the TS path, which fails open: enqueue is user-facing and
//! must not reject due to platform health, but the worker can safely
//! back off.

use std::sync::OnceLock;
use std::time::{SystemTime, UNIX_EPOCH};

use anyhow::{Context, Result};
use fred::clients::Client;
use fred::interfaces::{ClientLike, KeysInterface};
use fred::types::config::Config as FredConfig;
use serde::Serialize;

/// Bucket TTL: 60s window + 5s slack so a check landing on a minute boundary
/// still sees a fresh TTL on the new bucket.
const BUCKET_TTL_SEC: i64 = 65;

/// Plan caps for `workflow_executions/minute`. `None` = unlimited.
///
/// Kept in lockstep with [`PLAN_CAPS_PER_MINUTE`] in the TS helper. Any
/// change here must be mirrored there — the two are the same product
/// surface, just two languages.
fn plan_cap(plan: &str) -> Option<u32> {
    match plan {
        "free" => Some(5),
        "starter" => Some(30),
        "pro" => Some(120),
        "business" => Some(600),
        "enterprise" => None,
        // Unknown plan IDs fall through to the strictest cap, matching the
        // TS `normalisePlan` behaviour. A misconfigured tenant must never
        // accidentally get a *higher* rate by sending garbage.
        _ => Some(5),
    }
}

/// Outcome of a single claim-side rate check.
///
/// `serde::Serialize` is derived so callers can include the result in
/// structured logs / telemetry without re-mapping the variants.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum RateCheck {
    /// Workspace is under its per-minute cap. Caller may claim the job.
    /// `remaining` is the tokens left in the *current* minute window after
    /// this check; on unlimited plans we surface `u32::MAX`.
    Allowed { remaining: u32 },
    /// Workspace is at or above its cap. Caller should skip the claim and
    /// retry after `retry_after_ms`.
    Denied { retry_after_ms: u64 },
}

impl RateCheck {
    /// True when the caller may proceed.
    pub fn is_allowed(self) -> bool {
        matches!(self, RateCheck::Allowed { .. })
    }
}

/// Per-workspace + per-plan claim-side rate check.
///
/// Uses a process-global fred [`Client`] lazily initialised from the
/// `REDIS_URL` env var (falling back to `redis://localhost:6379`).
///
/// Returns:
/// - `Ok(RateCheck::Allowed { remaining })` — claim may proceed.
/// - `Ok(RateCheck::Denied { retry_after_ms })` — claim must back off.
/// - `Err(_)` — Redis was unreachable. Dispatcher policy is to treat this
///   as "do not claim this tick"; see the module doc on fail-closed.
pub async fn check_claim_rate(workspace_id: &str, plan: &str) -> Result<RateCheck> {
    let client = redis_client().await?;
    check_claim_rate_with_client(client, workspace_id, plan).await
}

/// Same as [`check_claim_rate`] but takes a borrowed [`Client`] — useful
/// for tests and for callers that already own a connected fred client
/// (e.g. the dispatcher binary's main loop).
pub async fn check_claim_rate_with_client(
    client: &Client,
    workspace_id: &str,
    plan: &str,
) -> Result<RateCheck> {
    if workspace_id.is_empty() {
        anyhow::bail!("workspace_id must not be empty");
    }

    let cap = match plan_cap(plan) {
        // Unlimited plan — short-circuit without a Redis round-trip.
        None => {
            return Ok(RateCheck::Allowed {
                remaining: u32::MAX,
            });
        }
        Some(c) => c,
    };

    let minute = current_minute()?;
    let key = bucket_key(plan, workspace_id, minute);

    // INCR is atomic. Returns the new value as i64; we cast to u32 with
    // saturation in the (impossible-in-practice) case that someone manually
    // bumped the bucket past 2^32 before we observed it.
    let count: i64 = client
        .incr(&key)
        .await
        .context("INCR sabflow:rate bucket failed")?;

    if count == 1 {
        // First hit in this minute — set TTL. We deliberately do not set
        // EXPIRE on every call: doing so would push the bucket's expiry
        // forward indefinitely and bleed the window across minute boundaries
        // for hot workspaces.
        let _: bool = client
            .expire(&key, BUCKET_TTL_SEC, None)
            .await
            .context("EXPIRE sabflow:rate bucket failed")?;
    }

    if count as u64 > cap as u64 {
        // Roll back so the counter reflects accepted claims only. A failure
        // here is non-fatal — the bucket evicts itself in 65s anyway — so
        // we log and continue rather than propagate.
        if let Err(err) = client.decr::<i64, _>(&key).await {
            tracing::warn!(
                workspace_id,
                plan,
                error = %err,
                "failed to roll back rate-limit counter; bucket will over-count for one minute"
            );
        }
        return Ok(RateCheck::Denied {
            retry_after_ms: ms_until_next_minute()?,
        });
    }

    let remaining = cap.saturating_sub(count.max(0) as u32);
    Ok(RateCheck::Allowed { remaining })
}

// ── Internals ─────────────────────────────────────────────────────────

fn bucket_key(plan: &str, workspace_id: &str, minute: u64) -> String {
    format!("sabflow:rate:{plan}:{workspace_id}:{minute}")
}

fn current_minute() -> Result<u64> {
    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .context("system clock before UNIX epoch")?
        .as_secs();
    Ok(secs / 60)
}

fn ms_until_next_minute() -> Result<u64> {
    let now_ms = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .context("system clock before UNIX epoch")?
        .as_millis() as u64;
    let into_minute = now_ms % 60_000;
    Ok((60_000 - into_minute).max(1))
}

/// Process-global fred client, lazily initialised on first use.
///
/// Holding the connected `Client` in a `OnceLock` is the same pattern
/// `sabnode-db` uses inside its handle — fred's `Client` is `Arc` inside,
/// so subsequent callers reuse the existing connection task instead of
/// dialling Redis again.
static REDIS_CLIENT: OnceLock<Client> = OnceLock::new();

async fn redis_client() -> Result<&'static Client> {
    if let Some(client) = REDIS_CLIENT.get() {
        return Ok(client);
    }

    let url = std::env::var("REDIS_URL").unwrap_or_else(|_| "redis://localhost:6379".to_string());
    let config =
        FredConfig::from_url(&url).with_context(|| format!("parsing REDIS_URL `{url}`"))?;
    let client = Client::new(config, None, None, None);
    client
        .init()
        .await
        .context("initializing fred Redis client for sabflow-executor-queue")?;

    // Race-tolerant set: if another task already initialised, drop ours.
    // fred's connection task is owned by the Client we drop, which will
    // tear itself down when the handle is freed.
    match REDIS_CLIENT.set(client) {
        Ok(()) => Ok(REDIS_CLIENT.get().expect("just set above")),
        Err(_) => Ok(REDIS_CLIENT.get().expect("set by competing task")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bucket_key_matches_ts_shape() {
        // The TS side writes `sabflow:rate:<plan>:<workspaceId>:<minute>`.
        // Both helpers must agree byte-for-byte or the dispatcher will
        // observe a different counter than the enqueue path wrote to.
        assert_eq!(
            bucket_key("pro", "ws_123", 28_700_000),
            "sabflow:rate:pro:ws_123:28700000"
        );
    }

    #[test]
    fn unknown_plan_falls_back_to_free_cap() {
        assert_eq!(plan_cap("free"), Some(5));
        assert_eq!(plan_cap("nonsense"), Some(5));
        assert_eq!(plan_cap(""), Some(5));
    }

    #[test]
    fn enterprise_is_unlimited() {
        assert_eq!(plan_cap("enterprise"), None);
    }

    #[test]
    fn plan_caps_match_spec() {
        assert_eq!(plan_cap("starter"), Some(30));
        assert_eq!(plan_cap("pro"), Some(120));
        assert_eq!(plan_cap("business"), Some(600));
    }

    #[test]
    fn ms_until_next_minute_is_in_range() {
        let ms = ms_until_next_minute().unwrap();
        assert!(ms >= 1 && ms <= 60_000, "got {ms}");
    }

    #[test]
    fn rate_check_is_allowed_helper() {
        assert!(RateCheck::Allowed { remaining: 3 }.is_allowed());
        assert!(
            !RateCheck::Denied {
                retry_after_ms: 100
            }
            .is_allowed()
        );
    }
}
